src/rust/engine/fs/src/store.rs

```diff
@@ -384,21 +389,62 @@ impl Store {
   ///
   pub fn ensure_local_has_recursive_directory(&self, dir_digest: Digest) -> BoxFuture<(), String> {
     let store = self.clone();
+    let remote = if let Some(ref remote) = self.remote {
+      remote
+    } else {
+      // TODO: Remote isn't configured, implement proceeding without one
+      debug!("Remote isn't configured, proceeding without one");
+      return future::err("Cannot ensure local has dir without a remote".to_owned()).to_boxed();
+    };
+    let remote = remote.clone();
     self
       .load_directory(dir_digest)
       .and_then(move |directory_opt| {
         directory_opt.ok_or_else(|| format!("Could not read dir with digest {:?}", dir_digest))
       })
       .and_then(move |directory| {
-        // Traverse the files within directory
-        let file_futures = directory
-          .get_files()
-          .iter()
-          .map(|file_node| {
-            let file_digest = try_future!(file_node.get_digest().into());
-            store.load_bytes_with(EntryType::File, file_digest, |_| Ok(()), |_| Ok(()))
-          })
-          .collect::<Vec<_>>();
+        // Traverse the files within a directory and split them into batches or not depending on the
+        // size of the files and/or if Remote is configured.
+        let mut file_futures = Vec::new();
+        let mut digests_to_batch = Vec::new();
+        for file_node in directory.get_files() {
+          let file_digest = try_future!(file_node.get_digest().into());
+          // collect all the digests that will benefit from batching (small) only if remote
+          // is configured
+          if file_digest.1 <= MAX_DOWNLOAD_BATCH_SIZE_BYTES {
+            //&& remote.is_some() {
+            digests_to_batch.push(file_digest);
+          } else {
+            file_futures.push(store.load_bytes_with(
+              EntryType::File,
+              file_digest,
+              |_| Ok(()),
+              |_| Ok(()),
+            ));
+          }
+        }
+
+        let batched_futures = {
+          if digests_to_batch.len() > 0 {
+            let mut batches = vec![vec![]];
+            let mut current_batch_size = 0;
+            for req in digests_to_batch {
+              if current_batch_size + req.1 < MAX_DOWNLOAD_BATCH_SIZE_BYTES {
```

Comment:
This should probably loop over the existing batches, and probably shouldn't have a `current_batch_size`; right now it looks like you're always either pushing to the first batch or making a new one, but resetting current_batch_size for the first batch based on whether you created a new batch?